### DISCUSSION TITLE
"fill white" does not fill plain white

### DISCUSSION THREAD

**Discussion by marciano1** [ImageMagick/ImageMagick#2755] [Help] [answered]
Good evening,

For years I've been using convert and composite in a php script to automatically to text over images.
This is a part of the stamp:
![Screenshot from 2020-10-22 19-17-11](https://user-images.githubusercontent.com/5874277/96937428-78435900-149e-11eb-8931-2e624a35052b.jpg)
Today I realized that from some months ago the printed text is not white anymore.
![Screenshot from 2020-10-22 19-16-50](https://user-images.githubusercontent.com/5874277/96937431-79748600-149e-11eb-94d4-6badc874851c.jpg)

This is the code I've been using (print white text over a semi-transparent black background):
`$A1 = "convert -size {$wi}x{$he} xc:transparent \`
				`-draw \"fill black fill-opacity 0.3 rectangle {$a},0 {$wi},{$L} \`
					   `font DejaVu-Sans-Condensed font-size 13   \`
					  ` fill white  \`
					  ` translate {$b},{$L} rotate -90 text 0,0 'TEXT/{$n}' \" \`
				`draw_mvg.png";`

`exec("$A1") ;`

`$A2 = "composite -gravity Center draw_mvg.png {$foto} {$foto}";`

`exec("$A2");`

I don't get what happens.
Thank you

**Comment by snibgo** [comment]
What version of IM, on what platform?

Please show your commands without any variables. Otherwise we can only guess.

**Reply by marciano1** [reply]
IM 6.9.10-68 x86_64 2020-04-01 Centos 7.8.2003 php 7.2

This is to build a rectangle the same dimensions as the photo 440x250. At the top right corner a semi-transparent background 13x140 rectangle is created with white text on it.  Result is named draw_mvg.png

`convert -size 440x250 xc:transparent \`
`-draw \"fill black fill-opacity 0.3 rectangle 427,0 440,140 \`
`font DejaVu-Sans-Condensed font-size 13 \`
`fill white \`
`translate 438,140 rotate -90 text 0,0 '.com/67316' \" \`
`draw_mvg.png`

Here I compose draw_mvg.png and filename (the image file)
`composite -gravity Center draw_mvg.png filename filename`

Thank you

**Comment by snibgo** [comment]
I guess your final composite makes no difference, and the problem is visible in draw_mvg.png. Is that correct?

Personally, I find long complex draw operations difficult to understand, eg what opacity do you want for the text? I would split it like this:
```
convert -size 440x250 xc:transparent \
-draw "fill black fill-opacity 0.3 rectangle 427,0 440,140" \
-draw "font DejaVu-Sans-Condensed font-size 13 fill white translate 438,140 rotate -90 text 0,0 '.com/67316' " \
d.png
```

**Reply by marciano1** [reply]
Yes, I guess the problem is with draw_mvg.png.
I've been using that code for more than 10 years!
Over a transparent 440x250 rectangle I draw a 13x140 black rectangle 0.3 transparency from (440-13, 0; 440,140)
Also draw white text 2 pix margin (440-2)
Geometry is okay, the problem is that the text is not white, it seems it is below the semi-transparent rectangle

**Comment by snibgo** [comment, marked as answer]
Your command draws white text with fill-opacity 0.3 on a black background with fill-opacity 0.3. So some of that gray shows through the semi-transparent white.

I assume you don't want the white to be semi-transparent. I'm not sure how to write a single "-draw" operation that gives semi-transparent black but opaque white. So I would use two "-draw" operations.

**Reply by marciano1** [reply]
That worked.
I don't know why behavior has changed. I don't remember but maybe there was a change in the server, something like using imagick php module instead of the direct command, or something like that in the server. The script hasn't changed for years.
Thank you

**Reply by GeeMack** [reply]
You can change the opacity amount within a single "-draw" operation by adding another "fill-opacity" setting after drawing the rectangle.

          convert logo: -draw "fill black fill-opacity 0.3 rectangle 100,100 300,180 fill white fill-opacity 1 font-size 24 text 150,150 'TEST'" result.png
